syntax = "proto3";

package polaris.digitaltwin;

// Digital Twin gRPC service definition
service DigitalTwin {
    // Query current or historical system state
    rpc Query(QueryRequest) returns (QueryResponse);
    
    // Perform predictive simulation ("what-if" analysis)
    rpc Simulate(SimulationRequest) returns (SimulationResponse);
    
    // Perform root cause analysis and diagnosis
    rpc Diagnose(DiagnosisRequest) returns (DiagnosisResponse);
    
    // Manage Digital Twin lifecycle and introspection
    rpc Manage(ManagementRequest) returns (ManagementResponse);
}

// Query service messages
message QueryRequest {
    string query_id = 1;
    string query_type = 2;  // "current_state", "historical", "natural_language"
    string query_content = 3;
    map<string, string> parameters = 4;
    string timestamp = 5;
}

message QueryResponse {
    string query_id = 1;
    bool success = 2;
    string result = 3;
    double confidence = 4;
    string explanation = 5;
    string timestamp = 6;
    map<string, string> metadata = 7;
}

// Simulation service messages
message SimulationRequest {
    string simulation_id = 1;
    string simulation_type = 2;  // "forecast", "what_if", "scenario"
    repeated ControlAction actions = 3;
    int32 horizon_minutes = 4;
    map<string, string> parameters = 5;
    string timestamp = 6;
}

message ControlAction {
    string action_id = 1;
    string action_type = 2;
    string target = 3;
    map<string, string> params = 4;
    string priority = 5;
    double timeout = 6;
}

message SimulationResponse {
    string simulation_id = 1;
    bool success = 2;
    repeated PredictedState future_states = 3;
    SimulationMetrics metrics = 4;
    double confidence = 5;
    double uncertainty_lower = 6;
    double uncertainty_upper = 7;
    string explanation = 8;
    CostPerformanceReliability impact_estimates = 9;
    string timestamp = 10;
    map<string, string> metadata = 11;
}

message PredictedState {
    string timestamp = 1;
    map<string, double> metrics = 2;
    double confidence = 3;
    string description = 4;
}

message SimulationMetrics {
    double execution_time_sec = 1;
    int32 scenarios_processed = 2;
    double average_confidence = 3;
    map<string, double> custom_metrics = 4;
}

message CostPerformanceReliability {
    double cost_impact = 1;
    double performance_impact = 2;
    double reliability_impact = 3;
    string cost_currency = 4;
    string impact_description = 5;
}

// Diagnosis service messages
message DiagnosisRequest {
    string diagnosis_id = 1;
    string anomaly_description = 2;
    string timestamp = 3;
    map<string, string> context = 4;
}

message DiagnosisResponse {
    string diagnosis_id = 1;
    bool success = 2;
    repeated CausalHypothesis hypotheses = 3;
    string causal_chain = 4;
    double confidence = 5;
    string explanation = 6;
    repeated string supporting_evidence = 7;
    string timestamp = 8;
    map<string, string> metadata = 9;
}

message CausalHypothesis {
    string hypothesis = 1;
    double probability = 2;
    string reasoning = 3;
    repeated string evidence = 4;
    int32 rank = 5;
}

// Management service messages
message ManagementRequest {
    string request_id = 1;
    string operation = 2;  // "health_check", "reload_model", "get_metrics", "reset_state"
    map<string, string> parameters = 3;
    string timestamp = 4;
}

message ManagementResponse {
    string request_id = 1;
    bool success = 2;
    string result = 3;
    map<string, string> metrics = 4;
    HealthStatus health_status = 5;
    string timestamp = 6;
}

message HealthStatus {
    string status = 1;  // "healthy", "degraded", "unhealthy", "not_initialized"
    string last_check = 2;
    map<string, double> performance_metrics = 3;
    repeated string issues = 4;
    string model_type = 5;
    string model_version = 6;
}